### ISSUE TITLE
dtmin limit triggered with user-added tstops

### ISSUE THREAD

**Issue by tkoolen** [JuliaDiffEq/OrdinaryDiffEq.jl#273]
I don't think the following behavior is desired:

```julia
using OrdinaryDiffEq
du = [0.]
u0 = [0.]
f = (u, p, t) -> du
problem = ODEProblem(f, u0, (0., 1.))
integrator = init(problem, Tsit5())
add_tstop!(integrator, 0.5)
add_tstop!(integrator, 0.5 + eps(0.5))
solve!(integrator);
```
result:
```julia
WARNING: dt <= dtmin. Aborting. If you would like to force continuation with dt=dtmin, set force_dtmin=true
```

**Comment by ChrisRackauckas** [comment]
Does it have a use case?

**Comment by tkoolen** [comment]
Yeah, multiple `PeriodicCallback`s with periods that are multiples of each other. Rounding error makes them produce `tstop`s that are just a little bit different. Granted, maybe `PeriodicCallback`s should re-use an already-existing `tstop` if there is one very close to the desired time, but in general I think it's not great that the integration can be terminated just by adding a 'bad' `tstop`.

**Comment by ChrisRackauckas** [comment]
Yeah that makes sense. To get around the numerical issues of a really small `dt` it should probably just skip actually doing a step and go right to callbacks. It shouldn't be too hard to do that.

**Comment by tkoolen** [comment]
That may be a good solution, depending on what bad behavior the `dtmin` check is supposed to avoid.

* if the `dtmin` check is meant to avoid numerical issues due to integration with small time steps in general, then this is the right solution.
* if the `dtmin` check is meant as a proxy for detecting instability, I think this solution is not the right one. In this case, you'd really only want to do the check on `dt`s calculated by the integrator itself.

In any case, your proposal solves the problem of unnecessary extra evaluations of `f` in the multiple`PeriodicCallback`s I described, and it fixes my current use case.

**Comment by ChrisRackauckas** [comment]
It kind of does both. Technically other instability checks should kick in, but adaptivity some times prevents the real instability from occurring. So yeah, what it should do is be skipped when `dt` is forced small by a `tstop`, and in that case not do a step. That both prevents numerical and efficiency issues but also leaves the instability check in there.

**Comment by tkoolen** [comment]
In terms of implementation, there needs to be some epsilon below which a step will not be taken. What should that epsilon (or at least, its default value) be?

I must say I am a little worried that  given just `integrator.tstops`, it won't be possible to accurately discern between the case where a user adds a `tstop` that's close to another `tstop` and the case in which the `dt` computed by the integrator is small. I.e., there is no epsilon that separates the two cases with 100% accuracy. Again, I think your proposal is a lot better than the current situation, but I think the rock-solid solution would involve doing the `dtmin` check only on the `dt`s computed by the integrator.

**Comment by ChrisRackauckas** [comment]
>but I think the rock-solid solution would involve doing the dtmin check only on the dts computed by the integrator.

That is the proposed change.

**Comment by tkoolen** [comment]
OK, great. I must have misunderstood https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/273#issuecomment-373617112 then.
